libvast/vast/concept/convertible/data.hpp

```diff
@@ -0,0 +1,439 @@
+                                         "target variable [{}, {}]",
+                                         src, std::numeric_limits<To>::min(),
+                                         std::numeric_limits<To>::max()));
+    dst = detail::narrow_cast<To>(src);
+  }
+  return caf::none;
+}
+
+// Overload for integers.
+template <detail::signed_integral To>
+caf::error convert(const integer& src, To& dst, const integer_type&) {
+  if constexpr (sizeof(To) >= sizeof(integer::value)) {
+    dst = src.value;
+  } else {
+    if (src.value < std::numeric_limits<To>::min()
+        || src.value > std::numeric_limits<To>::max())
+      return caf::make_error(ec::convert_error,
+                             fmt::format(": {} can not be represented by the "
+                                         "target variable [{}, {}]",
+                                         src, std::numeric_limits<To>::min(),
+                                         std::numeric_limits<To>::max()));
+    dst = detail::narrow_cast<To>(src.value);
+  }
+  return caf::none;
+}
+
+// Overload for enums.
+// clang-format off
+template <class To>
+  requires std::is_enum_v<To>
+caf::error convert(const std::string& src, To& dst, const enumeration_type& t) {
+  auto i = std::find(t.fields.begin(), t.fields.end(), src);
+  if (i == t.fields.end())
+    return caf::make_error(ec::convert_error, ": {} is not a value of {}", src,
+        detail::pretty_type_name(dst));
+  dst = detail::narrow_cast<To>(std::distance(t.fields.begin(), i));
+  return caf::none;
+}
+// clang-format on
+
+// Overload for lists.
+template <detail::has_push_back To>
+caf::error convert(const list& src, To& dst, const list_type& t) {
+  size_t num = 0;
+  for (const auto& element : src) {
+    typename To::value_type v{};
+    if (auto err = convert(element, v, t.value_type))
+      return detail::prepend(std::move(err), "[{}]", num);
+    dst.push_back(std::move(v));
+    num++;
+  }
+  return caf::none;
+}
+
+// Overload for maps.
+// clang-format off
+template <detail::has_insert To>
+  requires requires {
+    typename To::key_type;
+    typename To::mapped_type;
+  }
+caf::error convert(const map& src, To& dst, const map_type& t) {
+  // TODO: Use structured bindings outside of the lambda once clang supports
+  // that.
+  for (const auto& x : src) {
+    auto err = [&] {
+      const auto& [data_key, data_value] = x;
+      typename To::key_type key{};
+      if (auto err = convert(data_key, key, t.key_type))
+        return err;
+      typename To::mapped_type value{};
+      if (auto err = convert(data_value, value, t.value_type))
+        return err;
+      return detail::insert_to_map(dst, std::move(key), std::move(value));
+    }();
+    if (err)
+      return detail::prepend(std::move(err), "{}", x.first);
+  }
+  return caf::none;
+}
+// clang-format on
+
+// Overload for record to map.
+// clang-format off
+template <detail::has_insert To>
+  requires requires {
```

Comment:
This on the other hand looks like a good contender for a named concept.

Reply:
The only problem that I have is that I can't come up with a concise name for a concept key-value container other than "map", but that can be too easily confused with `std::map`.

How about we change the namespace name to `concepts` to avoid these awkward name clashes?

Reply:
I think if we want to do this we should use a separate PR for that.

Reply:
I think we should change the namespace to `vast::concepts` to avoid the awkward `detail::<concept_name>` use at call sites anyway.

Reply:
> I think we should change the namespace to `vast::concepts`

I like that proposal, looks very readable to me.